src/parser/lexer.mll

```diff
@@ -692,24 +727,77 @@ and bad_markup_recovery start_offset input = parse
         (Parse_error.bad_markup ("{" ^ rest) ~suggestion);
       emit input (`Code_span text) ~start_offset}
 
-(* The second field of the metadata.
-   This rule keeps whitespaces and newlines in the 'metadata' field except the
-   ones just before the '['. *)
-and code_block_metadata_tail input = parse
- | (space_char+ as prefix)
-   ((space_char* (_ # space_char # ['['])+)+ as meta)
-   ((space_char* '[') as suffix)
-    {
-      let meta =
-        with_location_adjustments ~adjust_start_by:prefix ~adjust_end_by:suffix (fun _ -> Loc.at) input meta
-      in
-      `Ok (Some meta)
-    }
-  | (newline | horizontal_space)* '['
-    { `Ok None }
-  | _ as c
-    { `Invalid_char c }
+(* Based on OCaml's parsing/lexer.mll
+   We're missing a bunch of cases here, and can add them
+   if necessary. Using the missing cases will cause a warning *)
+and string input = parse
+ | '\"'
+   { let result = Buffer.contents string_buffer in
+     Buffer.clear string_buffer;
+     result }
+ | '\\' newline [' ' '\t']*
+   { string input lexbuf }
+ | '\\' (['\\' '\'' '\"' 'n' 't' 'b' 'r' ' '] as c)
+   { Buffer.add_char string_buffer
+       (match c with
+        | '\\' -> '\\'
+        | '\'' -> '\''
+        | '\"' -> '\"'
+        | 'n' -> '\n'
+        | 't' -> '\t'
+        | 'b' -> '\b'
+        | 'r' -> '\r'
+        | ' ' -> ' '
+        | _ -> assert false);
+     string input lexbuf }
+  | '\\' ['0'-'9'] ['0'-'9'] ['0'-'9']
+    { Buffer.add_char string_buffer (char_for_decimal_code input lexbuf 1);
+      string input lexbuf }
+  | '\\' (_ as c)
+    { warning input (Parse_error.should_not_be_escaped c);
+      Buffer.add_char string_buffer c;
+      string input lexbuf }
   | eof
+    { warning input Parse_error.truncated_string;
+      Buffer.contents string_buffer }
+  | (_ as c)
+    { Buffer.add_char string_buffer c;
+      string input lexbuf }
+  
+and code_block_metadata_atom input = parse
+ | '"'
+   {
+    let start_offset = Lexing.lexeme_start input.lexbuf + 1 in
+    Buffer.clear string_buffer;
+    let s = string input lexbuf in
+    with_location_adjustments ~start_offset ~adjust_end_by:"\"" (fun _ -> Loc.at) input s }
+ | (tag_unquoted_atom as value)
+   { with_location_adjustments (fun _ -> Loc.at) input value }
+ | (_ as c) 
+   { warning input (Parse_error.code_block_tag_invalid_char c);
+     with_location_adjustments (fun _ -> Loc.at) input "" }
```

Comment:
Nitpick: It's quite hard to read what values `c` can take, but in fact it can only be "=". I don't know if that is more unreadable to match on "=" or on "_"... but I think at least a comment would help!

```suggestion
 | (tag_unquoted_atom as value)
   { with_location_adjustments (fun _ -> Loc.at) input value }
 | ('=' as c) 
   { warning input (Parse_error.code_block_tag_invalid_char c);
     with_location_adjustments (fun _ -> Loc.at) input "" }
```

```suggestion
 | (tag_unquoted_atom as value)
   { with_location_adjustments (fun _ -> Loc.at) input value }
 | (_ as c)   (* At this point, c can only be "=" *) 
   { warning input (Parse_error.code_block_tag_invalid_char c);
     with_location_adjustments (fun _ -> Loc.at) input "" }
```

Reply:
I think I prefer the explicit '=' version.